### PR TITLE
Support swap

### PR DESCRIPTION
**Pull request by Tougee**

### CODE REVIEW

File: lib/ui/page/swap.dart [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+    assert(supportedAssets.length > 1);
+    final sourceAsset = useState(supportedAssets[0]);
+    final destAsset = useState(supportedAssets[1]);
+    final routeData = useState<RouteData?>(null);
+    final sourceTextController = useTextEditingController();
+    final destTextController = useTextEditingController();
+    final sourceFocusNode = useFocusNode(debugLabel: 'source input');
+    final slippage = calcSlippage(routeData.value);
+    final slippageDisplay = displaySlippage(slippage);
+
```
Comment (YeungKC) [inline lib/ui/page/swap.dart:96]: Avoid executed in every build, should use useMemoized.

File: lib/ui/page/swap_detail.dart [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+    final order = useState<Order?>(null);
+    useEffect(() {
+      var canceled = false;
+      scheduleMicrotask(() async {
+        while (!canceled) {
+          try {
+            final traceOrder = (await MixSwap.client.getOrder(traceId)).data;
+            sourceId.value = traceOrder.payAssetUuid;
+            destId.value = traceOrder.receiveAssetUuid;
```
Comment (YeungKC) [inline lib/ui/page/swap_detail.dart:45]: Would you extract code snippet under from condition to be a hook widget?

**Review by YeungKC — CHANGES_REQUESTED** [review]

File: lib/ui/page/swap_detail.dart (hunk 2 above)
Comment (Tougee) [reply lib/ui/page/swap_detail.dart]: I've extracted this part as a function for now.

File: lib/ui/page/asset_detail.dart [hunk 3]
```diff
@@ -324,13 +324,7 @@ class _HeaderButtonBar extends StatelessWidget {
...
           HeaderButton.text(
             text: context.l10n.swap,
-            onTap: () {
-              ScaffoldMessenger.of(context)
-                ..hideCurrentSnackBar()
-                ..showSnackBar(
-                  SnackBar(content: Text(context.l10n.comingSoon)),
-                );
-            },
+            onTap: () => context.push(swapPath),
```
Comment (boyan01) [inline lib/ui/page/asset_detail.dart:327]: navigate to swap from asset_detail ,but didn't use current asset as paramater?

File: lib/ui/widget/mixin_appbar.dart [hunk 4]
```diff
@@ -50,8 +53,9 @@ class MixinAppBar extends StatelessWidget with PreferredSizeWidget {
           color: Colors.white,
         ),
         backgroundColor: backgroundColor ?? context.theme.accent,
+        bottom: bottom,
       );
 
   @override
-  Size get preferredSize => const Size.fromHeight(48);
+  Size get preferredSize => Size.fromHeight(bottom != null ? 96 : 48);
```
Comment (boyan01) [inline lib/ui/widget/mixin_appbar.dart:60]: how about `48 + bottom?.preferredSize.height ?? 0`

**Review by YeungKC — APPROVED** [review]

File: lib/ui/page/swap.dart [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+  }) : super(key: key);
+
+  final Client swapClient;
+  final List<AssetResult> supportedAssets;
+
+  @override
+  Widget build(BuildContext context) {
+    assert(supportedAssets.length > 1);
+    final sourceAsset = useState(supportedAssets[0]);
+    final destAsset = useState(supportedAssets[1]);
```
[the remark this answers is not in the document]
Comment (Tougee) [reply lib/ui/page/swap.dart:88]: good point, I missed this.

**Review by boyan01 — APPROVED** [review]

**Pull request merged**